tribe-ext-pdf-tickets/index.php

```diff
@@ -0,0 +1,633 @@
+
+		if ( empty( $html ) ) {
+			return $successful;
+		}
+
+		$file_name = $this->get_pdf_path( $attendee_id );
+
+		if ( empty( $file_name ) ) {
+			return $successful;
+		}
+
+		if ( file_exists( $file_name ) ) {
+			$successful = true;
+		} else {
+			$successful = $this->output_pdf( $html, $file_name );
+		}
+
+		/**
+		 * Action hook after the PDF ticket gets created.
+		 *
+		 * Might be useful if you want the PDF file added to the Media Library
+		 * via wp_insert_attachment(), for example.
+		 *
+		 * @var $ticket_class
+		 * @var $event_id
+		 * @var $attendee_id
+		 * @var $file_name
+		 */
+		do_action( 'tribe_ext_pdf_tickets_uploaded_pdf', $ticket_class, $event_id, $attendee_id, $file_name );
+
+		/**
+		 * Filter to disable PDF email attachments, either entirely (just pass
+		 * false) or per event, attendee, ticket type, or some other logic.
+		 *
+		 * @var $email
+		 * @var $ticket_class
+		 * @var $event_id
+		 * @var $attendee_id
+		 * @var $file_name
+		 */
+		$email = (bool) apply_filters( 'tribe_ext_pdf_tickets_attach_to_email', $email, $ticket_class, $event_id, $attendee_id, $file_name );
+
+		if (
+			true === $successful
+			&& true === $email
+		) {
+			$this->attachments_array[] = $file_name;
+
+			if ( 'Tribe__Tickets__RSVP' === $ticket_class ) {
+				add_filter( 'tribe_rsvp_email_attachments', array( $this, 'email_attach_pdf' ) );
+			} elseif ( 'Tribe__Tickets_Plus__Commerce__WooCommerce__Main' === $ticket_class ) {
+				add_filter( 'tribe_tickets_plus_woo_email_attachments', array( $this, 'email_attach_pdf', ) );
+			} elseif ( 'Tribe__Tickets_Plus__Commerce__EDD__Main' === $ticket_class ) {
+				add_filter( 'edd_ticket_receipt_attachments', array( $this, 'email_attach_pdf' ) );
+			} else {
+				// unknown ticket type so no emailing to do
+			}
+		}
+
+		return $successful;
+	}
+
+	/**
+	 * Attach the queued PDF(s) to the ticket email.
+	 *
+	 * RSVP, Woo, and EDD filters all just pass an attachments array so we can
+	 * get away with a single, simple function here.
+	 *
+	 * @param $attachments
+	 *
+	 * @return array
+	 */
+	public function email_attach_pdf( $attachments ) {
+		$attachments = array_merge( $attachments, $this->attachments_array );
+
+		// just a sanity check
+		$attachments = array_unique( $attachments );
+
+		return $attachments;
+	}
+
+	/**
+	 * Create the HTML link markup (a href) for a PDF Ticket file.
+	 *
+	 * @param $attendee_id
+	 *
+	 * @return string
+	 */
+	protected function ticket_link( $attendee_id ) {
+		$text = __( 'PDF Ticket', 'tribe-extension' );
+
+		/**
+		 * Filter to control the link target for Attendees Report links.
+		 *
+		 * @param $target
+		 */
+		$target = apply_filters( 'tribe_ext_pdf_tickets_link_target', '_blank' );
+
+		$output = sprintf(
+			'<a href="%s"',
+			esc_attr( $this->get_pdf_url( $attendee_id ) )
+		);
+
+		if ( ! empty( $target ) ) {
+			$output .= sprintf( ' target="%s"',
+				esc_attr( $target )
+			);
+		}
+
+		$output .= sprintf(
```

Comment:
Just a thought, if you wanted to prettify the downloaded filename we could do so here using the `<a download="">` attribute. 

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-download

Reply:
> the user will be prompted to save it as a local file.
That's not what we want.